File: server/collector-engine/strategia_collector/connectors/pdf_connector.py
import logging
import requests
from io import BytesIO

from strategia_collector.connectors.connector_base import BaseConnector, CollectedData

logger = logging.getLogger(__name__)


class PdfConnector(BaseConnector):
    TIMEOUT = 30

    async def fetch(self, url: str) -> list[CollectedData]:
        logger.info(f"Fetching PDF: {url}")

        try:
            resp = requests.get(url, timeout=self.TIMEOUT)
            resp.raise_for_status()

            import PyPDF2
            reader = PyPDF2.PdfReader(BytesIO(resp.content))
            text = ""
            for page in reader.pages:
                text += page.extract_text() or ""

            items = [CollectedData(
                source_url=url,
                source_type="PDF",
                title=url.split("/")[-1] or "PDF Document",
                description=text[:200],
                content=text,
                content_raw=text,
                published_at=None,
            )]

            logger.info(f"PDF: extracted {len(text)} chars")
            return items

        except Exception as e:
            logger.error(f"PDF fetch failed for {url}: {e}")
            raise

File: server/backend/src/modules/collection-engine/connectors/pdf.connector.ts
import { Injectable, Logger } from '@nestjs/common';
import * as fs from 'fs';
import * as path from 'path';
import { IConnector, CollectedData } from './connector.interface';

@Injectable()
export class PdfConnectorService implements IConnector {
  private readonly logger = new Logger(PdfConnectorService.name);
  private readonly TIMEOUT = 30000;

  async fetch(url: string): Promise<CollectedData[]> {
    try {
      let pdfBuffer: Buffer;
      if (url.startsWith('http')) {
        const response = await fetch(url, { signal: AbortSignal.timeout(this.TIMEOUT) });
        pdfBuffer = Buffer.from(await response.arrayBuffer());
      } else {
        pdfBuffer = fs.readFileSync(url);
      }

      const pdfParse = require('pdf-parse');
      const data = await pdfParse(pdfBuffer);

      this.logger.log(`PDF: extracted ${data.text.length} chars from ${path.basename(url)}`);

      return [{
        sourceUrl: url,
        sourceType: 'PDF',
        title: data.info?.Title || path.basename(url),
        description: data.text.substring(0, 200),
        content: data.text,
        contentRaw: data.text,
        publishedAt: data.info?.CreationDate ? new Date(data.info.CreationDate) : undefined,
      }];
    } catch (error: any) {
      this.logger.error(`PDF fetch failed for ${url}: ${error.message}`);
      return [];
    }
  }
}
